src/algebra/cubic_solution.lean

```diff
@@ -0,0 +1,570 @@
+    exact div_ne_zero ne_zero (nonzero_of_invertible 27) h_iszero,
+  assumption,
+end
+
+lemma multiply_out_cubic [field F] (x a b c : F) :
+  (x + a) * (x + b) * (x + c) =
+  x * x * x + (a + b + c) * x * x + (a * b + a * c + b * c) * x + (a * b * c) :=
+begin
+  ring,
+end
+
+/--
+The three solutions obtained by using different cube roots in the
+depressed_cubic_solution formula factorize the depressed cubic.
+--/
+lemma depressed_cubic_factorize [field F] [invertible (2 : F)] [invertible (3 : F)]
+       (c d x sqrt cubrt rt3_i : F) [invertible c]
+       /- TODO: These ne_zero arguments are redundant with the invertibles,
+          but I haven't been able to get the proof to work properly without them.-/
+      (two_ne_zero : (2 : F) ≠ (0 : F)) (three_ne_zero : (3 : F) ≠ (0 : F))
+      (c_ne_zero : c ≠ (0 : F)) :
+  (sqrt * sqrt) = (depressed_cubic_solution_sqrt c d) ->
+  (cubrt * cubrt * cubrt) = (depressed_cubic_solution_cube_root d sqrt) ->
+  (rt3_i * rt3_i = (-3)) ->
+    (x - depressed_cubic_solution c cubrt) *
+    (x - depressed_cubic_solution c (cubrt * cube_root_unity_a rt3_i)) *
+    (x - depressed_cubic_solution c (cubrt * cube_root_unity_b rt3_i))
+    =
+    x * x * x + c * x + d :=
+begin
+  intros h_sqrt h_cubrt h_rt3i,
+  have cubrt_ne_zero : cubrt ≠ 0,
+    intros cubrt_zero, rw cubrt_zero at h_cubrt, rw mul_zero at h_cubrt,
+    symmetry' at h_cubrt, unfold depressed_cubic_solution_cube_root at h_cubrt,
+    rw add_comm at h_cubrt, rw <- sub_eq_add_neg at h_cubrt,
+    exact formula_den_inv c d sqrt h_sqrt h_cubrt,
+  repeat {rw sub_eq_add_neg}, rw multiply_out_cubic,
+  have x_squ : (-depressed_cubic_solution c cubrt + -depressed_cubic_solution c (cubrt * cube_root_unity_a rt3_i) + -depressed_cubic_solution c (cubrt * cube_root_unity_b rt3_i))
+                = 0,
+    unfold depressed_cubic_solution, field_simp,
+    repeat {rw <- mul_assoc}, rw div_mul_eq_div_mul_one_div _ _ (cube_root_unity_a _),
+    rw cubrt_unity_a_reciprocal,
+    rw div_mul_eq_div_mul_one_div _ _ (cube_root_unity_b _),
+    rw cubrt_unity_b_reciprocal,
+    field_simp,
+    unfold cube_root_unity_a, unfold cube_root_unity_b, field_simp, ring,
+    repeat {assumption},
+  repeat {rw <- add_assoc}, repeat {rw mul_comm _ cubrt}, rw x_squ, clear x_squ,
+  repeat {rw zero_mul}, rw add_zero,
+  have x_term : (-depressed_cubic_solution c cubrt *
+                      -depressed_cubic_solution c (cubrt * cube_root_unity_a rt3_i) +
+                  -depressed_cubic_solution c cubrt *
+                      -depressed_cubic_solution c (cubrt * cube_root_unity_b rt3_i) +
+                  -depressed_cubic_solution c (cubrt * cube_root_unity_a rt3_i) *
+                      -depressed_cubic_solution c (cubrt * cube_root_unity_b rt3_i))
+                = c,
+    unfold depressed_cubic_solution, field_simp,
+    repeat {rw <- mul_assoc}, rw div_mul_eq_div_mul_one_div _ _ (cube_root_unity_a _),
```

Comment:
Mathlib style is to use `←` not `<-`; would you midn making this replacement throughout?
```suggestion
    repeat {rw ← mul_assoc}, rw div_mul_eq_div_mul_one_div _ _ (cube_root_unity_a _),
```